pennylane/measurements/probs.py

```diff
@@ -108,5 +111,180 @@ def circuit():
                 "Cannot specify the wires to probs if an observable is "
                 "provided. The wires for probs will be determined directly from the observable."
             )
-        return MeasurementProcess(Probability, wires=qml.wires.Wires(wires))
-    return MeasurementProcess(Probability, obs=op)
+        return _Probability(Probability, wires=qml.wires.Wires(wires))
+    return _Probability(Probability, obs=op)
+
+
+# TODO: Make public when removing the ObservableReturnTypes enum
+class _Probability(SampleMeasurement, StateMeasurement):
+    """Measurement process that computes the probability of each computational basis state."""
+
+    def process_samples(
+        self,
+        samples: Sequence[complex],
+        wire_order: Wires,
+        shot_range: Tuple[int] = None,
+        bin_size: int = None,
+    ):
+        wire_map = dict(zip(wire_order, range(len(wire_order))))
+        mapped_wires = [wire_map[w] for w in self.wires]
+        if shot_range is not None:
+            # Indexing corresponds to: (potential broadcasting, shots, wires). Note that the last
+            # colon (:) is required because shots is the second-to-last axis and the
+            # Ellipsis (...) otherwise would take up broadcasting and shots axes.
+            samples = samples[..., slice(*shot_range), :]
+
+        if mapped_wires:
+            # if wires are provided, then we only return samples from those wires
+            samples = samples[..., mapped_wires]
+
+        num_wires = qml.math.shape(samples)[-1]
+        # convert samples from a list of 0, 1 integers, to base 10 representation
+        powers_of_two = 2 ** qml.math.arange(num_wires)[::-1]
+        indices = samples @ powers_of_two
+
+        # `samples` typically has two axes ((shots, wires)) but can also have three with
+        # broadcasting ((batch_size, shots, wires)) so that we simply read out the batch_size.
+        batch_size = samples.shape[0] if qml.math.ndim(samples) == 3 else None
+        dim = 2**num_wires
+        # count the basis state occurrences, and construct the probability vector
+        if bin_size is not None:
+            num_bins = samples.shape[-2] // bin_size
+            return self._count_binned_samples(indices, batch_size, dim, bin_size, num_bins)
+        return self._count_unbinned_samples(indices, batch_size, dim)
+
+    def process_state(self, state: Sequence[complex], wire_order: Wires):
+        num_wires = len(wire_order)
+        dim = 2**num_wires
+        # Compute batch_size
+        expected_shape = [2] * num_wires
+        expected_size = dim
+        size = qml.math.size(state)
+        batch_size = (
+            size // expected_size
+            if qml.math.ndim(state) > len(expected_shape) or size > expected_size
+            else None
+        )
```

Comment:
Similar concern here. Most of this is copied over from the `default.qubit` device, but this same logic wouldn't work for other devices. Even the `default.mixed` device has a different way of computing the probabilities. To be honest, I don't think we should have an implementation that only works for `default.qubit`, since other devices would then have to override it, which is a big reason for this refactor in the first place.

Again, no clear answer here. This is open for discussion.

Reply:
As mentioned in the other comment, we agreed with @albi3ro to use the code from `QubitDevice` because it's the simplest measurement processing. But we will also allow devices to have custom implementations.